### PR TITLE
[Impeller] rectangle packer actually packs.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/148251

Work towards https://github.com/flutter/flutter/issues/138798

The rectangle packer was only filling up the top and right edge of each rectangle, making the atlas super inefficient.

### CODE REVIEW

File: impeller/typographer/typographer_unittests.cc [hunk 1]
```diff
@@ -325,46 +325,6 @@ TEST_P(TypographerTest, RectanglePackerAddsNonoverlapingRectangles) {
   ASSERT_EQ(packer->PercentFull(), 0);
 }
 
-TEST_P(TypographerTest, GlyphAtlasTextureIsRecycledWhenContentsAreRecreated) {
```
Comment (jonahwilliams) [inline impeller/typographer/typographer_unittests.cc:328]: This test really depends on the exact atlas sizing and how efficient the packing is, making it unstable. I'm removing it for now.

File: impeller/typographer/rectangle_packer.cc [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
 #include <algorithm>
 #include <vector>
+#include "fml/logging.h"
```
Comment (chinmaygarde) [inline impeller/typographer/rectangle_packer.cc:9]: Nit: flutter/fml/logging.h

File: impeller/typographer/rectangle_packer.cc [hunk 3]
```diff
@@ -29,7 +30,7 @@ class SkylineRectanglePacker final : public RectanglePacker {
 
   bool AddRect(int w, int h, IPoint16* loc) final;
 
-  Scalar PercentFull() const final {
+  float PercentFull() const final {
```
Comment (chinmaygarde) [inline impeller/typographer/rectangle_packer.cc:33]: Scalars are floats already.

File: impeller/typographer/rectangle_packer.cc [hunk 4]
```diff
@@ -50,10 +51,14 @@ class SkylineRectanglePacker final : public RectanglePacker {
   // the skyline segments >= 'skylineIndex'? If so, return true and fill in
   // 'y' with the y-location at which it fits (the x location is pulled from
   // 'skylineIndex's segment.
-  bool rectangleFits(int skylineIndex, int width, int height, int* y) const;
+  bool rectangleFits(size_t skylineIndex, int width, int height, int* y) const;
```
Comment (chinmaygarde) [inline impeller/typographer/rectangle_packer.cc:54]: Here and elsewhere, use the `PascalCase` per the style guide?

File: impeller/typographer/rectangle_packer.cc [hunk 5]
```diff
@@ -129,14 +135,14 @@ void SkylineRectanglePacker::addSkylineLevel(int skylineIndex,
...
+  skyline_.insert(skyline_.begin() + skylineIndex, newSegment);
 
   FML_DCHECK(newSegment.x_ + newSegment.width_ <= this->width());
   FML_DCHECK(newSegment.y_ <= this->height());
 
   // delete width of the new skyline segment from following ones
-  for (int i = skylineIndex + 1; i < (int)skyline_.size(); ++i) {
-    // The new segment subsumes all or part of skyline_[i]
+  for (auto i = skylineIndex + 1; i < skyline_.size(); ++i) {
+    // The new segment subsumes all or part of fSkyline[i]
```
Comment (chinmaygarde) [inline impeller/typographer/rectangle_packer.cc:145]: If fSkyline referring to the right ivar?

File: impeller/typographer/rectangle_packer.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/typographer/rectangle_packer.cc]: Fixed

File: impeller/typographer/rectangle_packer.cc (hunk 4 above)
Comment (jonahwilliams) [reply impeller/typographer/rectangle_packer.cc]: Fixed

File: impeller/typographer/rectangle_packer.cc (hunk 5 above)
Comment (jonahwilliams) [reply impeller/typographer/rectangle_packer.cc]: Fixed

File: impeller/typographer/rectangle_packer.cc [hunk 6]
```diff
@@ -67,9 +73,9 @@ bool SkylineRectanglePacker::AddRect(int width, int height, IPoint16* loc) {
   int bestX = 0;
   int bestY = this->height() + 1;
   int bestIndex = -1;
-  for (int i = 0; i < (int)skyline_.size(); ++i) {
+  for (auto i = 0u; i < skyline_.size(); ++i) {
     int y;
-    if (this->rectangleFits(i, width, height, &y)) {
+    if (this->RectangleFits(i, width, height, &y)) {
```
Comment (chinmaygarde) [inline impeller/typographer/rectangle_packer.cc:78]: `this->` is redundant.

**Review by matanlurey — APPROVED** [review]

File: impeller/typographer/rectangle_packer.cc [hunk 7]
```diff
@@ -158,11 +165,11 @@ void SkylineRectanglePacker::addSkylineLevel(int skylineIndex,
...
   }
 
-  // merge skyline_s
-  for (int i = 0; i < ((int)skyline_.size()) - 1; ++i) {
+  // merge skylines
+  for (auto i = 0u; i < skyline_.size() - 1; ++i) {
     if (skyline_[i].y_ == skyline_[i + 1].y_) {
       skyline_[i].width_ += skyline_[i + 1].width_;
-      skyline_.erase(std::next(skyline_.begin(), i));
+      skyline_.erase(skyline_.begin() + i + 1);
```
Comment (chinmaygarde) [inline impeller/typographer/rectangle_packer.cc:172]: I always have a tough time reading the stuff that returns an iterator you must use to modify the container.

But, this is an off by one error right? That is `std::next(begin(), i) + 1` would have been the same thing?

**Review by chinmaygarde — APPROVED** [review]

File: impeller/typographer/rectangle_packer.cc (hunk 7 above)
Comment (jonahwilliams) [reply impeller/typographer/rectangle_packer.cc:172]: Yeah, looking at the diff I think this was the bug.

Comment (chinmaygarde) [reply impeller/typographer/rectangle_packer.cc:172]: Thanks. I was wondering what was different.

File: impeller/typographer/rectangle_packer.cc (hunk 6 above)
Comment (jonahwilliams) [reply impeller/typographer/rectangle_packer.cc]: Done

**Pull request merged**